pkg/reconciler/pipeline/dag/dag.go

```diff
@@ -128,26 +128,21 @@ func linkPipelineTasks(prev *Node, next *Node) error {
 	// Check if we are adding cycles.
 	visited := map[string]bool{prev.Task.HashKey(): true, next.Task.HashKey(): true}
 	path := []string{next.Task.HashKey(), prev.Task.HashKey()}
-	if err := visit(next.Task.HashKey(), prev.Prev, path, visited); err != nil {
+	if err := visit(prev.Prev, path, visited); err != nil {
 		return fmt.Errorf("cycle detected: %w", err)
 	}
 	next.Prev = append(next.Prev, prev)
 	prev.Next = append(prev.Next, next)
 	return nil
 }
 
-func visit(currentName string, nodes []*Node, path []string, visited map[string]bool) error {
-	var sb strings.Builder
+func visit(nodes []*Node, path []string, visited map[string]bool) error {
 	for _, n := range nodes {
 		path = append(path, n.Task.HashKey())
 		if _, ok := visited[n.Task.HashKey()]; ok {
 			return errors.New(getVisitedPath(path))
 		}
-		sb.WriteString(currentName)
-		sb.WriteByte('.')
-		sb.WriteString(n.Task.HashKey())
-		visited[sb.String()] = true
```

Comment:
I'm a bit confused :S We no longer add anything to `visited` except for the two linked nodes (line 129).  Is that right?

Reply:
If that's right, do we need the `visisted` map at all?

Reply:
> I'm a bit confused :S We no longer add anything to visited except for the two linked nodes (line 129). Is that right?

we no longer add anything to visited. `visited` map does not need any update within the `visit` function. `linkPipelineTasks` function adds two pipelineTasks in the visited map before creating a link between the two.

The overall logic is:

`Build` function adds all the pipelineTasks from the `tasks` section in the `Graph`.  And creates a map of a `pipelineTask -> a list of parents (resource and order dependencies)`.

For the unit test I have added, the dependency [map](https://github.com/tektoncd/pipeline/blob/master/pkg/reconciler/pipeline/dag/dag.go#L79) looks like, e.g. node `g` has a parent node `f`, etc:

```
(map[string][]string) (len=7) {
 (string) (len=1) "g": ([]string) (len=1 cap=1) {
  (string) (len=1) "f"
 },
 (string) (len=1) "a": ([]string) (len=1 cap=1) {
  (string) (len=1) "e"
 },
 (string) (len=1) "b": ([]string) (len=1 cap=1) {
  (string) (len=1) "a"
 },
 (string) (len=1) "c": ([]string) (len=1 cap=1) {
  (string) (len=1) "a"
 },
 (string) (len=1) "d": ([]string) (len=2 cap=2) {
  (string) (len=1) "b",
  (string) (len=1) "c"
 },
 (string) (len=1) "e": ([]string) (len=1 cap=1) {
  (string) (len=1) "d"
 },
 (string) (len=1) "f": ([]string) (len=1 cap=1) {
  (string) (len=1) "d"
 }
}
``` 

Now, its iterating over this map to add a link between a `pipelineTask` and its parent node (one parent at a time). Before adding a link, it checks for any cycles.

Here, pt is the pipelineTask `b` and `previousTask` is actually parent `a`.

https://github.com/tektoncd/pipeline/blob/473e3f3cc74b215f8c22cb007fcd7413f93f3917/pkg/reconciler/pipeline/dag/dag.go#L84

Now, `addLink` with `prev` (parent) set to `a` and `next` set to `b`  calls `linkPipelineTasks`.  `linkPipelineTasks` created `visited` map with `a` and `b` and call `visit` with `b` and parents of `a` (`prev.Prev`), in the unit test, parent of `a` is `e`. `visit` checks if `e` exists in the `visited`, which is not `true` so now recursively call with `e` parents. `e` does not have any parents at this point of time and creates path `e -> a -> b` which is valid, no cycles. Update the `graph` setting `b.Prev=a` and `a.Next=b`.

Cycle is detected by iteratively checking parent and its parent until there are no parents in the `visited` map.

> If that's right, do we need the visisted map at all?

`visited` map is needed to check if any of the parent match with what we are visiting right now. 

https://github.com/tektoncd/pipeline/blob/473e3f3cc74b215f8c22cb007fcd7413f93f3917/pkg/reconciler/pipeline/dag/dag.go#L143